oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,449 @@
+   * @throws IOException if the credential cannot be created from the stream.
+   */
+  public static ExternalAccountCredentials fromStream(InputStream credentialsStream)
+      throws IOException {
+    return fromStream(credentialsStream, OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+  }
+
+  /**
+   * Returns credentials defined by a JSON file stream.
+   *
+   * <p>This will either return a IdentityPoolCredentials or AwsCredentials.
+   *
+   * @param credentialsStream the stream with the credential definition.
+   * @param transportFactory the HTTP transport factory used to create the transport to get access
+   *     tokens.
+   * @return the credential defined by the credentialsStream.
+   * @throws IOException if the credential cannot be created from the stream.
+   */
+  public static ExternalAccountCredentials fromStream(
+      InputStream credentialsStream, HttpTransportFactory transportFactory) throws IOException {
+    checkNotNull(credentialsStream);
+    checkNotNull(transportFactory);
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    GenericJson fileContents =
+        parser.parseAndClose(credentialsStream, OAuth2Utils.UTF_8, GenericJson.class);
+    return fromJson(fileContents, transportFactory);
+  }
+
+  /**
+   * Returns external account credentials defined by JSON using the format generated by gCloud.
+   *
+   * @param json a map from the JSON representing the credentials.
+   * @param transportFactory HTTP transport factory, creates the transport used to get access
+   *     tokens.
+   * @return the credentials defined by the JSON.
+   */
+  public static ExternalAccountCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) {
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
+    String subjectTokenType = (String) json.get("subject_token_type");
+    String tokenUrl = (String) json.get("token_url");
+    String tokenInfoUrl = (String) json.get("token_info_url");
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+
+    Map<String, Object> credentialSourceMap = (Map<String, Object>) json.get("credential_source");
+
+    // Optional params.
+    String clientId = json.containsKey("client_id") ? (String) json.get("client_id") : null;
+    String clientSecret =
+        json.containsKey("client_secret") ? (String) json.get("client_secret") : null;
+    String quotaProjectId =
+        json.containsKey("quota_project_id") ? (String) json.get("quota_project_id") : null;
+
+    if (isAwsCredential(credentialSourceMap)) {
+      // TODO(lsirac) return AWS credential here.
+    }
+    return new IdentityPoolCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        tokenInfoUrl,
+        new IdentityPoolCredentialSource(credentialSourceMap),
```

Comment:
Should knowledge of this inner class belong only to the `IdentityPoolCredentials`?